src/icmp.rs

```diff
@@ -0,0 +1,395 @@
+            },
+            0x04 => match c {
+                0x00 => Self::SourceQuench,
+                _ => Self::Other(raw),
+            },
+            0x05 => match c {
+                0x00 => Self::Redirect(Redirect::Network),
+                0x01 => Self::Redirect(Redirect::Host),
+                0x02 => Self::Redirect(Redirect::TosAndNetwork),
+                0x03 => Self::Redirect(Redirect::TosAndHost),
+                _ => Self::Other(raw),
+            },
+            0x07 => Self::Reserved,
+            0x08 => Self::EchoRequest,
+            0x09 => Self::RouterAdvertisment,
+            0x0A => Self::RouterSolicication,
+            0x0B => match c {
+                0x00 => Self::TimeExceeded(TimeExceeded::TTL),
+                0x01 => Self::TimeExceeded(TimeExceeded::FragmentReassembly),
+                _ => Self::Other(raw),
+            },
+            0x0C => match c {
+                0x00 => Self::ParameterProblem(ParameterProblem::Pointer),
+                0x01 => Self::ParameterProblem(ParameterProblem::MissingRequiredOption),
+                0x02 => Self::ParameterProblem(ParameterProblem::BadLength),
+                _ => Self::Other(raw),
+            },
+            0x0D => Self::Timestamp,
+            0x0E => Self::TimestampReply,
+            0x2A => Self::ExtendedEchoRequest,
+            0x2B => match c {
+                0x00 => Self::ExtendedEchoReply(ExtendedEchoReply::NoError),
+                0x01 => Self::ExtendedEchoReply(ExtendedEchoReply::MalformedQuery),
+                0x02 => Self::ExtendedEchoReply(ExtendedEchoReply::NoSuchInterface),
+                0x03 => Self::ExtendedEchoReply(ExtendedEchoReply::NoSuchTableEntry),
+                0x04 => {
+                    Self::ExtendedEchoReply(ExtendedEchoReply::MupltipleInterfacesStatisfyQuery)
+                }
+                _ => Self::Other(raw),
+            },
+            _ => Self::Other(raw),
+        }
+    }
+}
+
+fn parse_icmp_code(input: &[u8]) -> IResult<&[u8], IcmpCode> {
+    let (input, code) = number::streaming::be_u16(input)?;
+
+    Ok((input, code.into()))
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Default)]
+#[repr(transparent)] pub struct IcmpPayloadPacket([u8; 8]);
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum IcmpData {
+    Unreachable {
+        nexthop_mtu: u16,
+        header: IPv4Header,
+        packet: IcmpPayloadPacket,
+    },
+    Redirect {
+        gateway: Ipv4Addr,
+        header: IPv4Header,
+        packet: IcmpPayloadPacket,
+    },
+    TimeExceeded {
+        header: IPv4Header,
+        packet: IcmpPayloadPacket,
+    },
+    None,
+}
+
+fn parse_ipv4_header_and_packet(input: &[u8]) -> IResult<&[u8], (IPv4Header, IcmpPayloadPacket)> {
+    let (input, header) = parse_ipv4_header(input)?;
+    if input.len() < 8 {
+        return Err(Err::Incomplete(Needed::Size(8 - input.len())));
+    }
```

Comment:
`take()` take care of that

Reply:
Not in the same way. The `take()` always return 8 in this case.

Reply:
that not what the doc say. I just test it it return the correct amount missing.

Reply:
I realised that they've changed this behaviour from v5 to v6.
Good catch, thanks.
I'm fixing it since we are moving to nom v6 in the neighbouring PR anyway.

Reply:
OOO indeed you totally right my bad

Reply:
maybe we should merge the 6 PR before

Reply:
The downside that my tests are failing when I remove this check.

Reply:
I'd say that it's better to move to nom 6 first.
Let's wait for https://github.com/bestouff/pktparse-rs/pull/11 to be merged

Reply:
I'll merge 5->6 right now

Reply:
Done

Reply:
Thanks.
I've migrated to nom 6 in https://github.com/bestouff/pktparse-rs/pull/12/commits/06132b8c2e9cd9c2d79b5cc927ca6fc2e71b86b6